### PR TITLE
[ONNX] Fix scalar type promotion between fp16 tensor and fp32 scalar

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113780
* #113703
* #113697
* __->__ #113404

Fixes https://github.com/pytorch/pytorch/issues/104594.

The reason for the exporter behavior in original posted issue is explained as follows:
ONNX model track shape related computes that were done in pytorch by python
numbers as tensor computes. This is the only way for ONNX to track them properly
since ONNX only has tensor type, otherwise the computation result will be tracked
statically as constant, and the model won't work for another input that differs in shape.

Now for type promotion logic, scalars should be treated differently with tensors. 
Exporter mistook the shape related scalars as tensors in this case and incorrectly promoted.

This PR fixes the behavior and relaxes the criteria of scalar recognition. For floating point,
previously only a value from model initializer that has dtype torch.double and rank 0 is
treated as scalar. Now it is relaxed to any intermediate value, as well as for dtype torch.float.
Previous assumption was that python number is traced as torch.double dtype, which also
appears to be invalid anymore.

NOTE that this might introduce regression that a REAL 0-rank tensor is now being recognized as 
scalar. The downside is the model will drop in accuracy for these cases as certain computations 
will happen in lower precision data types.

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp [hunk 1]
```diff
@@ -161,6 +161,7 @@ static c10::optional<c10::ScalarType> InferExpectedScalarType(const Node* n) {
             at::typeMetaToScalarType(at::get_default_dtype());
         switch (scalar_type) {
           case at::kDouble:
+          case at::kFloat:
             // floating-point numbers wrapped as double tensors are
```
Comment (justinchuby) [inline torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp:165]:
```suggestion
            // floating-point numbers wrapped as float64/float32 tensors are
```

this?

File: torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp [hunk 2]
```diff
@@ -202,23 +203,17 @@ static c10::optional<c10::ScalarType> InferExpectedScalarType(const Node* n) {
...
-            }
-          }
         } else if (auto scalar_type = get_scalar_type(input)) {
-          typesFromTensors.emplace_back(*scalar_type);
+          auto tensor_type = input->type()->castRaw<TensorType>();
+          // get_scalar_type returns non-null value already guarantees
+          // that the input has a valid tensor_type.
+          TORCH_INTERNAL_ASSERT(nullptr != tensor_type);
+          auto rank = tensor_type->dim();
+          if (rank && rank.value() == 0) {
```
Comment (justinchuby) [inline torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp:212]: Maybe add the pr description as comment here?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -7101,6 +7101,14 @@ def forward(self, fp16, fp32):
         fp32 = Tensor([1.5])
         self.run_test(CatModel(), (fp16, fp32))
 
+        class DoNotUpcastModel(torch.nn.Module):
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7104]: Can this be a separate test case? I would explain a little more on what it is testing in code.

**Review by justinchuby — APPROVED** [review]
Thanks!

**Comment by huydhn** [comment]
Somehow, the test is still failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/39ca5a3226331428465a84d53d5b50dfb4406cfe, so I have disabled it here https://github.com/pytorch/pytorch/issues/113526 instead.

**Comment by jeanschmidt** [comment]
it is breaking [a test (X linux-focal-py3.8-clang10-onnx / test (default, 2, 2, linux.2xlarge))](https://hud.pytorch.org/pytorch/pytorch/commit/f2cd68102a56cd0427f25b748bbe3b463d43807b) on main.

I believe we should revert this

**Comment by BowenBao** [comment]
Hi @huydhn is there some way to find out how many of the tests were passing vs failing? I'm trying to see if this is a flaky test or if it is consistently failing.

**Comment by huydhn** [comment]
> Hi @huydhn is there some way to find out how many of the tests were passing vs failing? I'm trying to see if this is a flaky test or if it is consistently failing.

They are failing consistently from all the trunk commits that I see untill the commit was reverted https://hud.pytorch.org/hud/pytorch/pytorch/main/3?per_page=50
